inferno-framework/inferno-reference-server
# Initial token introspection with incomplete fields.

WIP DO NOT MERGE.  Intended to be used right now to support token introspection test development, and will need a few updates prior to merging.

# Summary
Adds token introspection capabilities per openid token introspection standard

## Testing guidance
This is intended to be run against the new token introspection tests under development in https://github.com/inferno-framework/smart-app-launch-test-kit/

There are a few things that are not yet complete, but i figured I'd let the tests under development surface those.

To use, I recommend:

* cloning this repository
* running `docker-compose up --build`.  Note, the first time, the db may be slow to launch because it is loading data, and you'll see an issue in the console and the site won't load. Kill the container and rerun `docker-compose up`, as the faster start to the db after it has been initialized should let the web server load
* confirm that http://localhost:8080/reference-server works and that there is a bullet stating support for 'Token Introspection'
* visit http://localhost:8080/reference-server/r4/.well-known/smart-configuration and there should be a URL for introspection_endpoint' (http://localhost:8080/reference-server/oauth/token/introspect)
*optiona: POST to this endpoint with no auth headers (I haven't added authorization here), and a body of `token=ABC`, and it should return json to the effect of `{'active': false}`

In the smart app launch test kit, assuming a 'ruby development setup', do the following:
* remove from `.env` and `.env.development` the `COMPOSE_PROFILES=keycloak` line -- i was having a little bit of trouble with this interferring with the reference server when running it at the same time, so lets just not start keycloak any more.
* `inferno services start` to start the background services (db, redis, etc)
* `inferno start` to start the application
* visit http://localhost:4567
* select STU2 of the smart app launch
* select 'Inferno Reference Server' preset
* Run 'standalone launch' tests
* change endpoint to http://localhost:8080/reference-server/r4
* click through and the tests should pass, except for TLS
* In test 1.1.01 (discovery), take a look at the output tab and grab the token introspection url from there (will be http://localhost:8080/reference-server/oauth/token/introspect)
* run the token introspection test
* appropriately split up the URL so that the current inputs work (note how we will want a single introspection url input insteead of having them split up in two inputs later)
* turn off 'include scope' thing for keycloak on the bottom
* run the group tests
* You'll get blocked by the fact that the token is not a JWK
* Remove a bunch of code in that test so that it really just verifying that the POST returns 'active = true'
* restart the inferno tests, it should now pass

This should allow us to much quicker iterate on the redesigned smart introspection tests.  Please let me know ASAP if any of these steps are giving you issue, running several docker-compose apps can be a little fiddly.  we might consider adding the reference server to the smart app launch docker-compose config as well as an optional thing so it is part of the 'test kit', but unfortunately that involves copying around a sql init file which is not good, so maybe not.
